Build failure: ceph

### Steps To Reproduce

Steps to reproduce the behavior:

``git checkout 9b19f5e77dd906cb52dade0b7bd280339d2a1f3d``
``nix build .#ceph``

### Build log

```
error: builder for '/nix/store/y6hik0qrizynz83x27a8jynkx5kzbfpj-ceph-18.2.0.drv' failed with exit code 2;
       last 10 log lines:
       > [ 74%] Building CXX object src/librbd/CMakeFiles/rbd_internal.dir/__/common/ContextCompletion.cc.o
       > [ 74%] Building CXX object src/librbd/CMakeFiles/rbd_internal.dir/crypto/luks/LUKSEncryptionFormat.cc.o
       > [ 74%] Building CXX object src/librbd/CMakeFiles/rbd_internal.dir/crypto/luks/FormatRequest.cc.o
       > [ 74%] Building CXX object src/librbd/CMakeFiles/rbd_internal.dir/crypto/luks/FlattenRequest.cc.o
       > [ 74%] Building CXX object src/librbd/CMakeFiles/rbd_internal.dir/crypto/luks/Header.cc.o
       > [ 74%] Building CXX object src/librbd/CMakeFiles/rbd_internal.dir/crypto/luks/LoadRequest.cc.o
       > [ 74%] Building CXX object src/librbd/CMakeFiles/rbd_internal.dir/crypto/luks/Magic.cc.o
       > [ 74%] Linking CXX static library ../../lib/librbd_internal.a
       > [ 74%] Built target rbd_internal
       > make: *** [Makefile:146: all] Error 2
       For full logs, run 'nix log /nix/store/y6hik0qrizynz83x27a8jynkx5kzbfpj-ceph-18.2.0.drv'.
```

https://hydra.nixos.org/job/nixpkgs/trunk/ceph.x86_64-linux

### Additional context

nixos-unstable `9b19f5e77dd906cb52dade0b7bd280339d2a1f3d`

Noticed when setting ``services.samba.package = pkgs.sambaFull;`` and rebuilding

### Notify maintainers
@adevress  @alexanderkjeldaas @johanot @krav 

<!--
Please @ people who are in the `meta.maintainers` list of the offending package or module.
If in doubt, check `git blame` for whoever last touched something.
-->

### Metadata

Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
❯ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 6.6.11, NixOS, 24.05 (Uakari), 24.05.20240113.9b19f5e`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.18.1`
 - channels(root): `"nixos-23.05"`
 - nixpkgs: `/etc/nix/path/nixpkgs`
```

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

The actual error:

https://cache.nixos.org/log/4i14y3ylpmagn2jfzi598ghw9fw2dza4-ceph-18.2.0.drv

```
/root/src/ceph/src/osd/SnapMapper.cc:331:0:   required from here
/nix/store/yc2pfwzxcxc6zadvfx76vfbpdg70m3xj-fmt-8.1.1-dev/include/fmt/ranges.h:490: error: call of overloaded ‘format_to(fmt::v8::appender&, const char [9], uint32_t&)’ is ambiguous
  490 |           out = format_to(out, "\\x{:02x}", escape.cp);
```

In colour so that one can actually make sense of C++ error messages:

![image](https://github.com/NixOS/nixpkgs/assets/399535/92af78d1-b980-49f9-9c8c-a2ea273cfbd4)

One potential problem here: `fmt-8.1.1-dev`

This is because of `all-packages.nix` having:

```
  inherit (callPackages ../tools/filesystems/ceph {
    lua = lua5_4;
    fmt = fmt_8;
  })
    ceph
    ceph-client;
```

* Switching to `fmt_9`:
* Removing `fmt = ` entirely (which swiches to `fmt_10`): **Fails** on new `fmt` related errors (see below)
* Removing `fmt = ` and `lua =`: **Fails** because Ceph requires lua >= 5.3 and nixpkgs's default is still 5.2

Build failure with `fmt_10`:

<details>

```
In file included from /nix/store/sgslg7d6dyr7962pc5dq6wl5g9p48bv2-fmt-10.1.1-dev/include/fmt/format.h:49,
                 from /build/ceph-18.2.0/src/msg/msg_types.h:21,
                 from /build/ceph-18.2.0/src/common/entity_name.h:22,
                 from /build/ceph-18.2.0/src/auth/Auth.h:19,
                 from /build/ceph-18.2.0/src/msg/async/crypto_onwire.h:22,
                 from /build/ceph-18.2.0/src/msg/async/frames_v2.h:6,
                 from /build/ceph-18.2.0/src/msg/async/frames_v2.cc:15:
/nix/store/sgslg7d6dyr7962pc5dq6wl5g9p48bv2-fmt-10.1.1-dev/include/fmt/core.h: In instantiation of 'constexpr decltype (ctx.begin()) fmt::v10::detail::parse_format_specs(ParseContext&) [with T = ceph_le<unsigned int>; ParseContext = compile_parse_context<char>; decltype (ctx.begin()) = const char*]':
/nix/store/sgslg7d6dyr7962pc5dq6wl5g9p48bv2-fmt-10.1.1-dev/include/fmt/core.h:2639:0:   required from here
/build/ceph-18.2.0/src/msg/async/frames_v2.cc:308:0:   in 'constexpr' expansion of 'fmt::v10::basic_format_string<char, unsigned int&, const ceph_le<unsigned int>&>("bad preamble crc calculated={} expected={}")'
/nix/store/sgslg7d6dyr7962pc5dq6wl5g9p48bv2-fmt-10.1.1-dev/include/fmt/core.h:2740:0:   in 'constexpr' expansion of 'fmt::v10::detail::parse_format_string<true, char, format_string_checker<char, unsigned int, ceph_le<unsigned int> > >(((fmt::v10::basic_format_string<char, unsigned int&, const ceph_le<unsigned int>&>*)this)->fmt::v10::basic_format_string<char, unsigned int&, const ceph_le<unsigned int>&>::str_, fmt::v10::detail::format_string_checker<char, unsigned int, ceph_le<unsigned int> >(fmt::v10::basic_string_view<char>(((const char*)s))))'
/nix/store/sgslg7d6dyr7962pc5dq6wl5g9p48bv2-fmt-10.1.1-dev/include/fmt/core.h:2523:0:   in 'constexpr' expansion of 'fmt::v10::detail::parse_replacement_field<char, format_string_checker<char, unsigned int, ceph_le<unsigned int> >&>(p, end, (* & handler))'
/nix/store/sgslg7d6dyr7962pc5dq6wl5g9p48bv2-fmt-10.1.1-dev/include/fmt/core.h:2457:0:   in 'constexpr' expansion of '(& handler)->fmt::v10::detail::format_string_checker<char, unsigned int, ceph_le<unsigned int> >::on_replacement_field((& handler)->fmt::v10::detail::format_string_checker<char, unsigned int, ceph_le<unsigned int> >::on_arg_id(), begin)'
/nix/store/sgslg7d6dyr7962pc5dq6wl5g9p48bv2-fmt-10.1.1-dev/include/fmt/core.h:2632:0:   in 'constexpr' expansion of '((fmt::v10::detail::format_string_checker<char, unsigned int, ceph_le<unsigned int> >*)this)->fmt::v10::detail::format_string_checker<char, unsigned int, ceph_le<unsigned int> >::on_format_specs(id, begin, begin)'
/nix/store/sgslg7d6dyr7962pc5dq6wl5g9p48bv2-fmt-10.1.1-dev/include/fmt/core.h:2548: error: 'fmt::v10::detail::type_is_unformattable_for<ceph_le<unsigned int>, char> _' has incomplete type
 2548 |     type_is_unformattable_for<T, char_type> _;
      | 
[ 11%] Linking CXX shared library ../../lib/libcls_hello.so
[ 11%] Building C object src/common/CMakeFiles/parse_secret_objs.dir/secret.c.o
[ 11%] Built target parse_secret_objs
[ 11%] Building C object src/tracing/CMakeFiles/osd_tp.dir/oprequest.c.o
[ 11%] Built target cls_hello
[ 11%] Building C object src/tracing/CMakeFiles/rados_tp.dir/librados.c.o
[ 11%] Building C object src/tracing/CMakeFiles/osd_tp.dir/osd.c.o
[ 12%] Building CXX object src/common/options/CMakeFiles/common-options-objs.dir/rbd-mirror_options.cc.o
/nix/store/sgslg7d6dyr7962pc5dq6wl5g9p48bv2-fmt-10.1.1-dev/include/fmt/core.h: In instantiation of 'constexpr fmt::v10::detail::value<Context> fmt::v10::detail::make_arg(T&) [with bool PACKED = true; Context = fmt::v10::basic_format_context<fmt::v10::appender, char>; T = const ceph_le<unsigned int>; typename std::enable_if<PACKED, int>::type <anonymous> = 0]':
/nix/store/sgslg7d6dyr7962pc5dq6wl5g9p48bv2-fmt-10.1.1-dev/include/fmt/core.h:1808:0:   required from 'constexpr fmt::v10::format_arg_store<Context, Args>::format_arg_store(T& ...) [with T = {unsigned int, const ceph_le<unsigned int>}; Context = fmt::v10::basic_format_context<fmt::v10::appender, char>; Args = {unsigned int, ceph_le<unsigned int>}]'
/nix/store/sgslg7d6dyr7962pc5dq6wl5g9p48bv2-fmt-10.1.1-dev/include/fmt/core.h:1826:0:   required from 'constexpr fmt::v10::format_arg_store<Context, typename std::remove_cv<typename std::remove_reference<_Args>::type>::type ...> fmt::v10::make_format_args(T& ...) [with Context = basic_format_context<appender, char>; T = {unsigned int, const ceph_le<unsigned int>}]'
/nix/store/sgslg7d6dyr7962pc5dq6wl5g9p48bv2-fmt-10.1.1-dev/include/fmt/core.h:2788:0:   required from 'std::string fmt::v10::format(format_string<T ...>, T&& ...) [with T = {unsigned int&, const ceph_le<unsigned int>&}; std::string = std::__cxx11::basic_string<char>; format_string<T ...> = basic_format_string<char, unsigned int&, const ceph_le<unsigned int>&>]'
/build/ceph-18.2.0/src/msg/async/frames_v2.cc:308:0:   required from here
/nix/store/sgslg7d6dyr7962pc5dq6wl5g9p48bv2-fmt-10.1.1-dev/include/fmt/core.h:1576: error: 'fmt::v10::detail::type_is_unformattable_for<const ceph_le<unsigned int>, char> _' has incomplete type
 1576 |     type_is_unformattable_for<T, typename Context::char_type> _;
      | 
/nix/store/sgslg7d6dyr7962pc5dq6wl5g9p48bv2-fmt-10.1.1-dev/include/fmt/core.h:1580: error: static assertion failed: Cannot format an argument. To make type T formattable provide a formatter<T> specialization: https://fmt.dev/latest/api.html#udt
 1580 |       formattable,
      | 
/nix/store/sgslg7d6dyr7962pc5dq6wl5g9p48bv2-fmt-10.1.1-dev/include/fmt/core.h:1580: note: 'formattable' evaluates to false
make[2]: *** [src/msg/CMakeFiles/common-msg-objs.dir/build.make:314: src/msg/CMakeFiles/common-msg-objs.dir/async/frames_v2.cc.o] Error 1
make[2]: *** Waiting for unfinished jobs....
```

</details>

I noticed Arch [has patches for `fmt` 10](https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=ceph&id=1aaec63d1fb688055d8b22ce50f1508a3fe6d3b7#n100).

But we can just use `fmt_9` to fix the build.

--- a/nixos/tests/ceph-multi-node.nix
+++ b/nixos/tests/ceph-multi-node.nix
@@ -185,6 +185,14 @@ let
     monA.succeed(
         "ceph osd pool create multi-node-test 32 32",
         "ceph osd pool ls | grep 'multi-node-test'",
+
+        # We need to enable an application on the pool, otherwise it will
+        # stay unhealthy in state POOL_APP_NOT_ENABLED.
+        # Creating a CephFS would do this automatically, but we haven't done that here.
+        # See: https://docs.ceph.com/en/reef/rados/operations/pools/#associating-a-pool-with-an-application
+        # We use the custom application name "nixos-test" for this.
+        "ceph osd pool application enable multi-node-test nixos-test",
+
         "ceph osd pool rename multi-node-test multi-node-other-test",
         "ceph osd pool ls | grep 'multi-node-other-test'",
     )

--- a/nixos/tests/ceph-single-node-bluestore.nix
+++ b/nixos/tests/ceph-single-node-bluestore.nix
@@ -145,6 +145,14 @@ let
     monA.succeed(
         "ceph osd pool create single-node-test 32 32",
         "ceph osd pool ls | grep 'single-node-test'",
+
+        # We need to enable an application on the pool, otherwise it will
+        # stay unhealthy in state POOL_APP_NOT_ENABLED.
+        # Creating a CephFS would do this automatically, but we haven't done that here.
+        # See: https://docs.ceph.com/en/reef/rados/operations/pools/#associating-a-pool-with-an-application
+        # We use the custom application name "nixos-test" for this.
+        "ceph osd pool application enable single-node-test nixos-test",
+
         "ceph osd pool rename single-node-test single-node-other-test",
         "ceph osd pool ls | grep 'single-node-other-test'",
     )

--- a/nixos/tests/ceph-single-node.nix
+++ b/nixos/tests/ceph-single-node.nix
@@ -145,6 +145,14 @@ let
     monA.succeed(
         "ceph osd pool create single-node-test 32 32",
         "ceph osd pool ls | grep 'single-node-test'",
+
+        # We need to enable an application on the pool, otherwise it will
+        # stay unhealthy in state POOL_APP_NOT_ENABLED.
+        # Creating a CephFS would do this automatically, but we haven't done that here.
+        # See: https://docs.ceph.com/en/reef/rados/operations/pools/#associating-a-pool-with-an-application
+        # We use the custom application name "nixos-test" for this.
+        "ceph osd pool application enable single-node-test nixos-test",
+
         "ceph osd pool rename single-node-test single-node-other-test",
         "ceph osd pool ls | grep 'single-node-other-test'",
     )
@@ -182,19 +190,16 @@ let
     monA.wait_until_succeeds("ceph -s | grep 'mgr: ${cfg.monA.name}(active,'")
     monA.wait_until_succeeds("ceph -s | grep 'HEALTH_OK'")
 
-    # This test has been commented out due to the upstream issue with pyo3
-    # that has broken this dashboard
-    # Reference: https://www.spinics.net/lists/ceph-users/msg77812.html
     # Enable the dashboard and recheck health
-    # monA.succeed(
-    #     "ceph mgr module enable dashboard",
-    #     "ceph config set mgr mgr/dashboard/ssl false",
-    #     # default is 8080 but it's better to be explicit
-    #     "ceph config set mgr mgr/dashboard/server_port 8080",
-    # )
-    # monA.wait_for_open_port(8080)
-    # monA.wait_until_succeeds("curl -q --fail http://localhost:8080")
-    # monA.wait_until_succeeds("ceph -s | grep 'HEALTH_OK'")
+    monA.succeed(
+        "ceph mgr module enable dashboard",
+        "ceph config set mgr mgr/dashboard/ssl false",
+        # default is 8080 but it's better to be explicit
+        "ceph config set mgr mgr/dashboard/server_port 8080",
+    )
+    monA.wait_for_open_port(8080)
+    monA.wait_until_succeeds("curl -q --fail http://localhost:8080")
+    monA.wait_until_succeeds("ceph -s | grep 'HEALTH_OK'")
   '';
 in {
   name = "basic-single-node-ceph-cluster";

--- a/pkgs/development/python-modules/cryptography/default.nix
+++ b/pkgs/development/python-modules/cryptography/default.nix
@@ -4,6 +4,7 @@
 , callPackage
 , cargo
 , cffi
+, cryptography-vectors ? (callPackage ./vectors.nix { })
 , fetchPypi
 , hypothesis
 , iso8601
@@ -24,9 +25,6 @@
 , setuptoolsRustBuildHook
 }:
 
-let
-  cryptography-vectors = callPackage ./vectors.nix { };
-in
 buildPythonPackage rec {
   pname = "cryptography";
   version = "41.0.7"; # Also update the hash in vectors.nix

--- a/pkgs/tools/filesystems/ceph/default.nix
+++ b/pkgs/tools/filesystems/ceph/default.nix
@@ -50,6 +50,9 @@
 , zlib
 , zstd
 
+# Dependencies of overridden Python dependencies, hopefully we can remove these soon.
+, rustPlatform
+
 # Optional Dependencies
 , curl ? null
 , expat ? null
@@ -166,7 +169,67 @@ let
 
   # Watch out for python <> boost compatibility
   python = python310.override {
-    packageOverrides = self: super: {
+    packageOverrides = self: super: let cryptographyOverrideVersion = "40.0.1"; in {
+      # Ceph does not support `cryptography` > 40 yet:
+      # * https://github.com/NixOS/nixpkgs/pull/281858#issuecomment-1899358602
+      # * Upstream issue: https://tracker.ceph.com/issues/63529
+      #   > Python Sub-Interpreter Model Used by ceph-mgr Incompatible With Python Modules Based on PyO3
+      #
+      # We pin the older `cryptography` 40 here;
+      # this also forces us to pin an older `pyopenssl` because the current one
+      # is not compatible with older `cryptography`, see:
+      #     https://github.com/pyca/pyopenssl/blob/d9752e44127ba36041b045417af8a0bf16ec4f1e/CHANGELOG.rst#2320-2023-05-30
+      cryptography = super.cryptography.overridePythonAttrs (old: rec {
+        version = cryptographyOverrideVersion;
+
+        src = fetchPypi {
+          inherit (old) pname;
+          version = cryptographyOverrideVersion;
+          hash = "sha256-KAPy+LHpX2FEGZJsfm9V2CivxhTKXtYVQ4d65mjMNHI=";
+        };
+
+        cargoDeps = rustPlatform.fetchCargoTarball {
+          inherit src;
+          sourceRoot = let cargoRoot = "src/rust"; in "${old.pname}-${cryptographyOverrideVersion}/${cargoRoot}";
+          name = "${old.pname}-${cryptographyOverrideVersion}";
+          hash = "sha256-gFfDTc2QWBWHBCycVH1dYlCsWQMVcRZfOBIau+njtDU=";
+        };
+
+        patches = (old.patches or []) ++ [
+          # Fix https://nvd.nist.gov/vuln/detail/CVE-2023-49083 which has no upstream backport.
+          # See https://github.com/pyca/cryptography/commit/f09c261ca10a31fe41b1262306db7f8f1da0e48a#diff-f5134bf8f3cf0a5cc8601df55e50697acc866c603a38caff98802bd8e17976c5R1893
+          ./python-cryptography-Cherry-pick-fix-for-CVE-2023-49083-on-cryptography-40.patch
+        ];
+
+        # Tests would require overriding `cryptography-vectors`, which is not currently
+        # possible/desired, see: https://github.com/NixOS/nixpkgs/pull/281858#pullrequestreview-1841421866
+        doCheck = false;
+      });
+
+      # This is the most recent version of `pyopenssl` that's still compatible with `cryptography` 40.
+      # See https://github.com/NixOS/nixpkgs/pull/281858#issuecomment-1899358602
+      pyopenssl = super.pyopenssl.overridePythonAttrs (old: rec {
+        version = "23.1.1";
+        src = fetchPypi {
+          pname = "pyOpenSSL";
+          inherit version;
+          hash = "sha256-hBSYub7GFiOxtsR+u8AjZ8B9YODhlfGXkIF/EMyNsLc=";
+        };
+      });
+
+      # Ceph does not support `kubernetes` >= 19, see:
+      #     https://github.com/NixOS/nixpkgs/pull/281858#issuecomment-1900324090
+      kubernetes = super.kubernetes.overridePythonAttrs (old: rec {
+        version = "18.20.0";
+        src = fetchFromGitHub {
+          owner = "kubernetes-client";
+          repo = "python";
+          rev = "v${version}";
+          sha256 = "1sawp62j7h0yksmg9jlv4ik9b9i1a1w9syywc9mv8x89wibf5ql1";
+          fetchSubmodules = true;
+        };
+      });
+
     };
   };
 

--- a/pkgs/tools/filesystems/ceph/python-cryptography-Cherry-pick-fix-for-CVE-2023-49083-on-cryptography-40.patch
+++ b/pkgs/tools/filesystems/ceph/python-cryptography-Cherry-pick-fix-for-CVE-2023-49083-on-cryptography-40.patch
@@ -1,0 +1,44 @@
+commit 08dbad8552ecca85d3e38072f94eb864b143d218
+Author: Niklas Hambüchen <mail@nh2.me>
+Date:   Mon Jan 22 12:52:06 2024 +0000
+
+    Cherry-pick fix for CVE-2023-49083
+    
+    Cherry-Picked-From: f09c261ca10a31fe41b1262306db7f8f1da0e48a
+    Cherry-Picked-By: Niklas Hambüchen <mail@nh2.me>
+
+diff --git a/src/cryptography/hazmat/backends/openssl/backend.py b/src/cryptography/hazmat/backends/openssl/backend.py
+index a3fe1bce4..58e7207c4 100644
+--- a/src/cryptography/hazmat/backends/openssl/backend.py
++++ b/src/cryptography/hazmat/backends/openssl/backend.py
+@@ -2383,9 +2383,12 @@ class Backend:
+                 _Reasons.UNSUPPORTED_SERIALIZATION,
+             )
+ 
++        certs: list[x509.Certificate] = []
++        if p7.d.sign == self._ffi.NULL:
++            return certs
++
+         sk_x509 = p7.d.sign.cert
+         num = self._lib.sk_X509_num(sk_x509)
+-        certs = []
+         for i in range(num):
+             x509 = self._lib.sk_X509_value(sk_x509, i)
+             self.openssl_assert(x509 != self._ffi.NULL)
+diff --git a/tests/hazmat/primitives/test_pkcs7.py b/tests/hazmat/primitives/test_pkcs7.py
+index 4e61c5ef5..d8170bfb9 100644
+--- a/tests/hazmat/primitives/test_pkcs7.py
++++ b/tests/hazmat/primitives/test_pkcs7.py
+@@ -89,6 +89,12 @@ class TestPKCS7Loading:
+                 mode="rb",
+             )
+ 
++    def test_load_pkcs7_empty_certificates(self, backend):
++        der = b"\x30\x0B\x06\x09\x2A\x86\x48\x86\xF7\x0D\x01\x07\x02"
++
++        certificates = pkcs7.load_der_pkcs7_certificates(der)
++        assert certificates == []
++
+ 
+ # We have no public verification API and won't be adding one until we get
+ # some requirements from users so this function exists to give us basic

--- a/pkgs/top-level/all-packages.nix
+++ b/pkgs/top-level/all-packages.nix
@@ -6809,8 +6809,14 @@ with pkgs;
 
   libceph = ceph.lib;
   inherit (callPackages ../tools/filesystems/ceph {
-    lua = lua5_4;
-    fmt = fmt_8;
+    lua = lua5_4; # Ceph currently requires >= 5.3
+
+    # To see which `fmt` version Ceph upstream recommends, check its `src/fmt` submodule.
+    #
+    # Ceph does not currently build with `fmt_10`; see https://github.com/NixOS/nixpkgs/issues/281027#issuecomment-1899128557
+    # If we want to switch for that before upstream fixes it, use this patch:
+    # https://github.com/NixOS/nixpkgs/pull/281858#issuecomment-1899648638
+    fmt = fmt_9;
   })
     ceph
     ceph-client;

